### ISSUE TITLE
Add (very specific) overload to Assert.Contains for Dictionary

### ISSUE THREAD

**Issue by ben-reilly** [xunit/xunit#1857]
Riffing off of #1620, I think there needs to be an overload Assert.Contains that specifically handles `Dictionary<TKey, TValue>`: #1620 introduced overloads for `IDictionary` and `IReadOnlyDictionary`, which creates ambiguity issues during compilation when using a `Dictionary` object like so:

```cs
var dictionary = new Dictionary<string, int>
{
    ["forty-two"] = 42,
};

Assert.Contains("forty-two", dictionary);
// Compiler error:
// The call is ambiguous between the following methods or properties:
// 'Assert.Contains<TKey, TValue>(TKey, IReadOnlyDictionary<TKey, TValue>)' and
// 'Assert.Contains<TKey, TValue>(TKey, IDictionary<TKey, TValue>)'
```

**Comment by bradwilson** [comment]
I'd like to see if we could reduce overloads rather than add them, if possible, as this would only solve it very specifically for one concrete implementation.

**Comment by ka-bezmaternykh** [comment]
Get same issue, are there any workarounds, except casting dictionary to generic interface? Cause it quite annoying and make code more verbose..

**Comment by meirkr** [comment]
I think it won't be understood to the person reading the tests, what is it about.
Does it refer to a key? to a value?

I would have proposed adding concrete methods for dictionary:
- Assert.ContainsKey(expectedKey, dictionary);
- Assert.ContainsValue(expectedValue, dictionary)

**But!**
Why not just using the existing methods?
If you like to assert they key exists: 
- Assert.Contains("forty-two", dictionary**.Keys**);
- Assert.Contains(42, dictionary**.Values**);

What do you think?

Would it close this request?

**Comment by rdnlsmith** [comment]
@meirkr I agree that `Assert.Contains("forty-two", dictionary.Keys);` is more clear.  However, it doesn't quite replace the dictionary overloads of `Assert.Contains()`, because those return the value for that key.

Suppose you had a dictionary of lists (no comment whether this is good practice):

```csharp
Dictionary<string, List<string>> casts = new()
{
    {
        "Star Trek",
        new List<string> { "William Shatner", "Leonard Nimoy", "DeForest Kelley", ... },
    },
    {
        "Star Trek: The Next Generation",
        new List<string> { "Patrick Stewart", "Jonathan Frakes", "Michael Dorn", ... },
    }
}
```

The dictionary `Assert.Contains()` makes it very convenient to do further assertions on each key.

```csharp
List<string> members = Assert.Contains("Star Trek", casts);
Assert.Collection(members, ... );
```

Obviously, it's not a big deal to assign `List<string> members = casts["Star Trek"];` after the initial assertion, but this return value was a key point in the original issue.

**Comment by safakgur** [comment]
> I'd like to see if we could reduce overloads rather than add them, if possible, as this would only solve it very specifically for one concrete implementation.

I completely agree with that.

Any proper dictionary implementation will have [IReadOnlyDictionary'2][ir-d] implemented. The big three ([Dictionary'2][d], [ConcurrentDictionary'2][cd], [ImmutableDictionary'2][id]) already do. So, we probably don't need the [IDictionary'2][i-d] overload at all.

[ir-d]: https://learn.microsoft.com/dotnet/api/system.collections.generic.ireadonlydictionary-2
[d]: https://learn.microsoft.com/dotnet/api/system.collections.generic.dictionary-2
[cd]: https://learn.microsoft.com/dotnet/api/system.collections.concurrent.concurrentdictionary-2
[id]: https://learn.microsoft.com/dotnet/api/system.collections.immutable.immutabledictionary-2
[i-d]: https://learn.microsoft.com/dotnet/api/system.collections.generic.idictionary-2

**Comment by bradwilson** [comment]
This can be closed, as we ended up adding the concrete types while fixing https://github.com/xunit/xunit/issues/2671

**Comment by safakgur** [comment]
> This can be closed, as we ended up adding the concrete types while fixing #2671

Feedback for future (maybe v3?):

If you'll follow the approach where concrete implementations are added, `ConcurrentDictionary` is another biggie in .NET that I tend to test often. I currently have to do `Assert.True(foo.TryGetValue(key, out var value), "Some key-not-found message")`.

But if you remove all `Contains` overloads for concrete dictionary types + `IDictionary`, and leave only the one targetting `IReadOnlyDictionary`, you could get the widest compatibility[^1] with the fewest number of methods, without the need to add new overloads for potential new concrete types.

[^1]: Widest in practice. I appreciate you currently support types that implement `IDictionary` without `IReadOnlyDictionary` as well, improving the types you cover in theory. But everything that implements `IDictionary` _should_ (and every built-in dictionary type _does_) implement `IReadOnlyDictionary`, and having just one `Contains` targeting it means there won't be any ambiguity warnings.

**Comment by bradwilson** [comment]
Yeah, the reason we didn't go down to just `IReadOnlyDictionary` is because of compatibility, and having just the interfaces causes the compiler to complain about type compatibility. It's no problem to add ConcurrentDictionary as well, so I'll re-open this.
